util/hash_bucket_histogram.go

```diff
@@ -0,0 +1,70 @@
+package util
+
+import (
+	"hash/fnv"
+	"reflect"
+	"sync/atomic"
+	"unsafe"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// HashBucketHistogramOpts are the options for making a HashBucketHistogram
+type HashBucketHistogramOpts struct {
+	prometheus.HistogramOpts
+	HashBuckets int
+}
+
+// HashBucketHistogram is used to track a histogram of per-bucket rates.
+//
+// For instance, I want to know that 50% of rows are getting X QPS or lower
+// and 99% are getting Y QPS of lower.  At first glance, this would involve
+// tracking write rate per row, and periodically sticking those numbers in
+// a histogram.  To make this fit in memory: instead of per-row, we keep
+// N buckets of counters and hash the key to a bucket.  Then every scrape
+// we update a histogram with the bucket values (and zero the buckets).
+type HashBucketHistogram interface {
+	prometheus.Metric
+	prometheus.Collector
+
+	Observe(string, uint32)
+}
+
+type hashBucketHistogram struct {
+	prometheus.Histogram
+	buckets []uint32
+}
+
+// NewHashBucketHistogram makes a new HashBucketHistogram
+func NewHashBucketHistogram(opts HashBucketHistogramOpts) HashBucketHistogram {
+	return &hashBucketHistogram{
+		Histogram: prometheus.NewHistogram(opts.HistogramOpts),
+		buckets:   make([]uint32, opts.HashBuckets, opts.HashBuckets),
+	}
+}
+
+// Collect implements prometheus.Metric
+func (h *hashBucketHistogram) Collect(c chan<- prometheus.Metric) {
+	for i := range h.buckets {
+		h.Histogram.Observe(float64(atomic.SwapUint32(&h.buckets[i], 0)))
```

Comment:
So what your histogram observes depends totally on your scrape rate, missed scrapes, and whether you scrape from multiple Prometheus servers in parallel. This is obviously funky and not normally recommended Prometheus metrics usage, but I guess you know that?

Also, the `xxx_count` will be off.